### PR TITLE
plutus-tx: support type families

### LINKED ISSUE #1434: Support type families
I thought we couldn't easily support type families, because that they were only solvable during typechecking and there wasn't a general "reduction" operation.

But there is! `FamInstEnv.normaliseType`. We can even get the required environments from `CoreM`, they're used in other simplifier passes.

We won't be able to handle "irreducible" type family applications, though.

### PR DESCRIPTION
**Pull request by michaelpj**
Fixes #1434.

There is one big restriction, which is that you can only use type families when they completely reduce. This means that e.g. `MyFamily a -> b` will probably not work. I think this will be okay for our use cases, however.

This gives us an immediate benefit, in that I can get rid of some type synonyms that were only there to avoid using type families. There's a bit of a diff in one of the contract PIR outputs, I think just because some stuff got reordered.

### CODE REVIEW

File: plutus-tx/test/Plugin/Data/Spec.hs [hunk 1]
```diff
@@ -226,3 +228,54 @@ sameEmptyRose = plc @"sameEmptyRose" (
...
+
+associatedParam :: CompiledCode Integer
+associatedParam = plc @"associatedParam" (paramId (Param 1) 1)
+
+-- Here we cannot reduce the type family
+{-# NOINLINE tfId #-}
+tfId :: forall a . a -> BasicClosed a -> BasicClosed a
+tfId _ x = x
+
+irreducible :: forall a . CompiledCode Integer
```
Comment (effectfully) [inline plutus-tx/test/Plugin/Data/Spec.hs]: What is the `a` for? And how does this line above:

```haskell
    , goldenPlcCatch "irreducible" irreducible
```

type check... I wanted to ask, but then

```haskell
Language.PlutusTx.Plugin:defer-errors
```

Comment (michaelpj) [reply plutus-tx/test/Plugin/Data/Spec.hs]: A leftover from when I was trying to put the quantification in here directly, which doesn't quite work. 

All the ones that use `*catch` are for errors.

**Comment by michaelpj** [comment]
I don't know about data families, I should try it.

**Comment by michaelpj** [comment]
Updated:
- Supports data families.
- I think I can make this more efficient by using `topNormaliseType`, which only normalizes the outermost stuff, but also normalizes away newtypes. If we're clever, this could simplify our handling of newtypes, but I need to think more about how to deal with recursive newtypes.

**Pull request merged**